src/main/java/sparksoniq/jsoniq/tuple/FlworKey.java

```diff
@@ -63,48 +63,84 @@ public FlworKey(List<Item> contents) {
+     * @return ResultIndexKeyTuple(result, index):
+     * result = -1 for smaller, 0 for equal, 1 for bigger (in ascending ordering)
+     * index = -1 equal
+     */
+    public ResultIndexKeyTuple compareWithFlworKey(FlworKey flworKey) {
         int result = 0;
-        for(Item currentItem: this.keyItems){
-            int index = this.keyItems.indexOf(currentItem);
-            Item comparisonItem = flworKey.getKeyItems().get(index);
-            if(currentItem == null || comparisonItem == null)
+        int sizeOfThisFlworKey = this.keyItems.size();
+        int sizeOfOtherFlworKey = flworKey.keyItems.size();
+
+        // iterate over every ordering expression of this flworkey
+        int index = 0;
+        while (index < sizeOfThisFlworKey) {
+            Item currentItem = this.keyItems.get(index);
+            Item comparisonItem;
+
+            // if the other Flworkey doesn't have the ordering expression result (it's empty)
+            // return 1 (bigger in ascending order) and the index of the current expression
+            if (index < sizeOfOtherFlworKey) {
+                comparisonItem = flworKey.getKeyItems().get(index);
+            } else {
                 return new ResultIndexKeyTuple(1, index);
+            }
 
+            // check for incorrect ordering inputs
             if (currentItem instanceof ArrayItem || currentItem instanceof ObjectItem ||
-                    comparisonItem instanceof ArrayItem || comparisonItem instanceof ObjectItem)
+                    comparisonItem instanceof ArrayItem || comparisonItem instanceof ObjectItem) {
                 throw new SparksoniqRuntimeException("Non atomic key not allowed");
-            else if(!currentItem.getClass().getSimpleName().equals(comparisonItem.getClass().getSimpleName())
-                    && (!Item.isNumeric(comparisonItem) || !Item.isNumeric(currentItem)))
+            } else if (!currentItem.getClass().getSimpleName().equals(comparisonItem.getClass().getSimpleName())
+                    && (!Item.isNumeric(comparisonItem) || !Item.isNumeric(currentItem))) {
                 throw new SparksoniqRuntimeException("Invalid sort key, different Item types");
-            else
+            } else {
                 result = Item.compareItems(currentItem, comparisonItem);
-
-            if(result != 0)
+            }
+
+            if (result == 0) {
+                index ++;
+                continue;
+            }
+            else {
+                // if comparison results in a difference, return the comparison result and ordering expression index
                 return new ResultIndexKeyTuple(result, index);
+            }
+        }
+        // execution reaches here when "this" Flworkey (can be empty) has matching ordering results  with the other Flworkey
+
+        // if ordering expressions results have equal size, Flworkeys are fully equal
+        if ( sizeOfThisFlworKey == sizeOfOtherFlworKey) {
```

Comment:
I don't think it ever happens that we compare two FLWOR keys with different sizes. You can simply throw a RuntimeException (to avoid a throw clause) if they don't match. This should simplify the code a lot.

Reply:
In the test cases I have debugged and observed this was not the case actually. Flworkeys of different sizes are compared particularly in the cases where one tuple doesn't have the necessary field. An example would be the following query:
```
jiqs$ for $i in (
   { "product" : "broiler", "store number" : 1, "quantity" : 20  },
   { "product" : "broiler", "store number" : 1, "quantity" : 30  },
   { "product" : "toaster", "store number" : 2  },
   { "product" : "toaster2", "store number2" : 2  })
order by $i."quantity" empty greatest
return $i
```
Flworkey of 1st and 2nd rows contain 1 entry
Flworkey of 3rd and 4th rows contain 0 entries.

Similarly:
```
for $i in (
   { "product" : "toaster", "store number" : 1  },
   { "product" : "broiler", "store number" : 1, "quantity" : 20  })
order by $i."store number", $i."quantity" descending empty greatest
return $i
```
1st row creates Flworkey with a single ordering field
2nd row creates a Flworkey with both ordering fields available.

Detection of empty fields are carried out using the index values returned from the Flworkey comparison in order to handle special empty field handling as greatest/least.

Please point out if you see any misconceptions. We can also discuss this further in this week's meeting.

Reply:
I see, thanks. Unfortunately, I am afraid that this implementation is not quite correct for two reasons:

1. The missing item may be a different positions in each row, which can "offset" the comparisons.

2. Whether an empty sequence is lowest or greatest depends on each variable, it is not for the entire order by clause.

Example:

for $i in (
   { "product" : "toaster", "store number" : 1  },
   { "product" : "broiler", "store number" : 1, "quantity" : 20  })
   { "product" : "broiler", "quantity" : 20  })
order by $i."store number" descending empty lowest, $i."quantity" descending empty greatest
return $i

I think the easiest way is simply to use Java null values in the FlowrKey map, and to handle them as part of the compareWithFlworKey() logics.

Also, this means that the return value can be an int, which would improve performance a lot comparing to creating a special object with two values.

Does it help?

Reply:
Yes this was very helpful. I have just pushed a new implementation covering the issues and the test case you've given above.

The indices of the object were used to determine which variable(expression) had the empty item and which ascending/descending selection it had. I was able to preserve the usage of such indices without the use of objects. It is implemented through a small trick with multiplication of integers.

The item comparison results in (or is simplified to, for the case of string comparison) +1/0/1. This value is multiplied by the indice of the expression/variable that the comparison result is determined by. Then, this value is extracted in the orderby closure to handle the empty items as specified in the query for each individual variable.